File: projecteuler/p5_smallest_multiple.py
#encoding=utf-8
"""
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
"""
#Runs in 0.000129 seconds
import time

startTime = time.clock()
primes = [2,3,5,7,11,13,17,19]
primeFactorsOfProduct = {}
finalProduct = 1

def primeFactors(n):
    factors = {}
    for prime in primes:
        while n%prime==0:
            n = n/prime
            if prime in factors:
                factors[prime] += 1
            else:
                factors[prime] = 1
    return factors

for x in xrange(2,21):
    for key in primeFactors(x):
        if key not in primeFactorsOfProduct:
            primeFactorsOfProduct[key] = primeFactors(x)[key]
        elif primeFactorsOfProduct[key] < primeFactors(x)[key]:
            primeFactorsOfProduct[key] = primeFactors(x)[key]

for key in primeFactorsOfProduct:
    for x in xrange(0,primeFactorsOfProduct[key]):
        finalProduct = finalProduct * key

print finalProduct, time.clock()-startTime

File: projecteuler-ruby/p5_smallest_multiple.rb
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

require 'set'

def smallestPositiveNumberDivisibleByAllNumbersUpTo(limit)
  allPrimeFactors = {}
  for number in 2..limit do
    primeFactors = primeFactors(number)
    primeFactors.each do |factor, amount|
      if amount > allPrimeFactors[factor].to_i
        allPrimeFactors[factor] = primeFactors[factor]
      end
    end
  end
  product = 1
  allPrimeFactors.each do |factor, amount|
    product *= factor ** amount
  end
  return product
end

def primeFactors(number)
  allPrimeFactors = []
  originalNumber = number
  primes = primeNumbersUpTo(number + 1)
  while number > 1 do
    for prime in primes do
      if number % prime == 0
        number /= prime
        allPrimeFactors << prime
        break
      end
    end    
  end
  allPrimeFactorsFrequencyHash = {}
  for factor in allPrimeFactors do
    allPrimeFactorsFrequencyHash[factor] = allPrimeFactorsFrequencyHash[factor].to_i + 1
  end
  return allPrimeFactorsFrequencyHash
end

def primeNumbersUpTo(limit)
  nonPrimeSet = Set[]
  primeArr = [2]
  currentPrime = 2
  
  while currentPrime < limit do
    currentMultiple = 2
    while currentMultiple * currentPrime <= limit do
      nonPrimeSet.add(currentMultiple * currentPrime)
      currentMultiple += 1
    end
    nextPrime = currentPrime + 1
    while nonPrimeSet.include?(nextPrime) do
      nextPrime += 1
    end
    if nextPrime < limit
      primeArr << nextPrime
    end
    currentPrime = nextPrime
  end
  return primeArr
end


puts smallestPositiveNumberDivisibleByAllNumbersUpTo(20)

